nin/frontend/app/scripts/directives/editor/GraphEditor.js

```diff
@@ -35,21 +35,35 @@ class GraphEditor extends React.Component {
     let depths = {};
     let seen = new Set();
 
+    depths.root = {
+      x: 0,
+      y: 0,
+      id: 'root',
+      counter: 0,
+    };
+    seen.add('root');
+
+    let counter = 1;
+
     function recurse(level, currentNodes) {
       let nextLevel = [];
       deepestLevel = Math.max(deepestLevel, level);
 
       for (let i = 0; i < currentNodes.length; i++) {
         let node = currentNodes[i];
-        seen.add(node.id);
-        depths[node.id] = {
-          x: level,
-          y: -i
-        };
 
         for (let child in node.inputs) {
           let source = node.inputs[child].source;
-          if (source && !seen.has(source.node.id)) {
+          if(source) {
+            depths[source.node.id] = {
+              x: level + 1,
+              y: 0,
+              id: source.node.id,
+              counter
+            };
+            counter++;
+          }
+          if (source) {
             seen.add(source.node.id);
```

Comment:
A thing i realize now is that we could drop the seen set entirely in favour of using the depths object (the keys would be the same as the contents of the set).